practice/rain.py

```diff
@@ -1,86 +1,214 @@
+    """Prints the day with the most rain, and how much rain fell in inches.
+
+    >>> output_for_max_rain_date('5','30-MAR-2016')
+    The day with the most rain was 30-MAR-2016, which had 5 inches of rain!
+    """
+    print('The day with the most rain was {}, which had {} inches of rain!'.format(date_with_most_rain, most_rain))
+
+
+def get_year(row):
+    """Calculates Grouping Key.
+
+    >>> get_year(['23', 'Mar', '2016'])
+    '2016'
+    """
+    return row[2]
+
+
+def get_year_to_rows(rows):
+    """Creates a dictionary of rows of data keyed to each year.
+
+    >>> get_year_to_rows([['23', 'Mar', '2016'], ['21', 'Feb', '2016']])
+    {'2016': [['23', 'Mar', '2016'], ['21', 'Feb', '2016']]}
+    """
+    year_to_rows = {
+        group_key: list(grouped_rows)
+        for group_key, grouped_rows
+        in groupby(rows, get_year)
+        }
+    return year_to_rows
+
+
+def sum_yearly_rainfall(year, year_to_daily_rain):
+    """Returns sum of yearly rainfall as an integer value
+
+    >>> sum_yearly_rainfall('2016', {['2016': [['1'], ['2']]]})
+    3
+    """
+    rain_fall_values = year_to_daily_rain[year]
+    daily_rain_values = [int(value) for value in rain_fall_values[3] if contains_numeral(value)]
+    total = sum(daily_rain_values)
+    return total
+
+
+def get_years_to_annual_rain(rows):
+    """Transforms a dictionary of years to data to years to total rain.
+
+    >>> get_years_to_annual_rain({[['2016']: [['1'], ['2']]], [['2017']: [['3'], ['4']]]})
+    {[['2016']: [[3]]], [['2017']: [[7]]]
+    """
+    year_to_rows = get_year_to_rows(rows)
+    years_to_annual_rain = {}
+    for year in year_to_rows:
+        annual_data = year_to_rows[year]
+        annual_rain_values = [int(value) for value in annual_data[3] if contains_numeral(value)]
+        total_annual_rain = sum(annual_rain_values)
+        year_to_annual_rain = {year: total_annual_rain}
+        years_to_annual_rain.update(year_to_annual_rain)
+    return years_to_annual_rain
+
+
+def get_year_with_most_rain(rows):
+    """Runs rows of data through functions to output year with most rainfall.
+
+    >>> get_year_with_most_rain(TEST_ROWS)
+    '2016'
+    """
+    years_to_annual_rain = get_years_to_annual_rain(rows)
+    year_with_most_rain = max(years_to_annual_rain.keys(), key=(lambda k: years_to_annual_rain[k]))
+    # found on Stack Overflow: http://tinyurl.com/jdjs5p6
+    return year_with_most_rain
+
+
+def output_for_year(year):
+    """ Prints the year with the most rain.
+
+    >>> output_for_year('2015')
+    The year with the most rain was 2015!
+    """
+    print('The year with the most rain was {}!'.format(year))
 
 
 def main():
-    text_lines = open_file()
+    """Main function."""
+    text_lines = open_file('rain_text/rain.txt')
     table_as_text_lines = cut_non_data(text_lines)
     rows_of_data = split_lines_to_rows(table_as_text_lines)
     columns_of_data = convert_rows_to_columns(rows_of_data)
```

Comment:
Names like `rows` and `columns` are hard to interpret when there's multiple data types in a matrix. Try coming up with more descriptive names throughout the program.